rfc/0015-data-generation.md

```diff
@@ -0,0 +1,132 @@
+lets pack developers define custonm workbenches of arbitrary size without
+needing to write a mod. There is no way of knowing what tables will be defined
+at compile time, so we can't possibly provide data satisfactorally. However,
+current runtime data generation tools do not cache their outputs, and therefore
+run every time the game is launched or resources are reloaded. For a mod with
+very few things to generate, this is not a problem, but when sizable mods like
+[Modern Industrialization](https://github.com/AztechMC/Modern-Industrialization)
+generate nearly all of their block and item models at reload time, this can
+result in significantly impacted load times, even if the data could be defined
+statically with no issue.
+
+## Explanation
+
+Data generation can easily be separated into *static* and *dynamic* generation.
+Static generation is not expected to change between launches or reloads; only
+between separate releases of a mod. Dynamic generation can change arbitrarily
+based on configuration, data, or other sources.
+
+Both static and dynamic generators will write to a `PackWriter` passed to the
+datagen method, which will manage loading of data and conversion to hard files.
+`PackWriter` contains methods to accept JSON objects, strings, or byte arrays
+to write to packs.
+
+### Static Generation
+
+Static data generation is performed in the `datagen` source set. `datagen` has
+access to `main`, as well as all dependencies accessible by `main`. Any methods
+with no return and accepting one `PackWriter` annotated by a `@DataGen`
+annotation will be invoked during build, with results written to hard viles.
+The `@DataGen` annotation requires an argument of `DataType.CLIENT_ASSETS` or
+`DataType.SERVER_DATA` to specify whether to write to `assets` or `data`. Any
+files added via datagen will be ignored if a file with that namespace and path
+in that type of data exists as a hard file in `src/main/resources`.
+
+### Dynamic Generation
+
+Dynamic data generation is performed through implementors of
+`DynamicDataGenerator`, which are registered to `ClientDataGenManager` for
+resource packs and `ServerDataGenManager` for data packs. A 
+`DynamicDataGenerator` defines both a method invoked during resource reload and
+a method that returns a state object as JSON. The `onReload` method will pass
+path and content strings to a `PackWriter`, which will place those files in a
+`quilt_generated` folder in the base game directory. The state object will
+be written as well, and compared to whenever a resource reload is triggered.
+If the current state and the saved state are different, the resources will be
+regenerated, allowing for reliable caching.
+
+## Drawbacks
+
+- Getting multiple source sets, especially arbitrary source sets, to play nice
+with IDEs can be challenging. This would require work in Quilt Gradle to ensure
+that source sets are executed on any build, whether it's to test or to release.
```

Comment:
I had made several (3 in total) data generators for MC modding, and I can whole-heartedly tell you that IntelliJ IDEA and Gradle just do not work with multiple source sets/modules and annotation processors; I often need to regenerate IDEA modules since the mapping from Gradle subprojects/source sets often get messed up, my annotation processor sometimes refuses to work in the test mod, and the main reason I stopped developing that data generator is that Mixins just don't apply *at all*, and so Fabric API crashes spetacularly.
Maybe this is a complex issue that is related to how IDEA, its Gradle plugin and Fabric/Quilt Loom exactly works, but to my experience it's not really pleasant.

Reply:
![](https://cdn.discordapp.com/emojis/756157231858188348.png?v=1)